File: crc4.cpp
#include <iostream>
#include <vector>
#include <iomanip>

// CRC-4 "klasse" für einfaches Hin und Zurück
class CRC4 {
    uint8_t polynomial = 0x13; // 0b10011
public:
    CRC4() = default;

    uint8_t compute(const std::vector<uint8_t>& data) const {
        uint8_t crc = 0;
        for (auto byte : data) {
            crc ^= (byte << 4); // auf obere 4 Bits shiften
            for (int i = 0; i < 8; ++i) {
                crc = (crc & 0x80) ? ((crc << 1) ^ polynomial) : (crc << 1);
            }
        }
        return crc >> 4; // nur 4 Bits
    }

    bool verify(const std::vector<uint8_t>& data, uint8_t crc) const {
        return compute(data) == crc;
    }
};

int main() {
    CRC4 crc4;
    std::vector<uint8_t> data = {0xAB, 0xCD};
    uint8_t crc = crc4.compute(data);

    std::cout << "Daten: ";
    for(auto b : data) std::cout << std::hex << (int)b << " ";
    std::cout << "\nCRC-4: " << std::hex << (int)crc << std::endl;

    std::cout << "Verifikation: "
              << (crc4.verify(data, crc) ? "OK" : "FEHLER") 
              << std::endl;
}

File: crc4.py
#!/usr/bin/env python3
class CRC4:
    def __init__(self, polynomial=0x13):
        self.polynomial = polynomial  # 0b10011

    def compute(self, data):
        crc = 0
        for byte in data:
            crc ^= (byte << 4)  # auf obere 4 Bits shiften
            for _ in range(8):
                crc = ((crc << 1) ^ self.polynomial) if (crc & 0x80) else (crc << 1)
                crc &= 0xFF  # nur 8 Bit
        return crc >> 4

    def verify(self, data, crc):
        return self.compute(data) == crc

data = [0xAB, 0xCD]
crc4 = CRC4()
crc = crc4.compute(data)

print("Daten:", [hex(b) for b in data])
print("CRC-4:", hex(crc))
print("Verifikation:", "OK" if crc4.verify(data, crc) else "FEHLER")
